### PR TITLE
netboot: the beginnings

### PR DESCRIPTION
**Pull request by hugelgupf**

### CODE REVIEW

File: cmds/pxeboot/pxeboot.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+func Netboot() error {
+	ifnames, err := netlink.LinkList()
+	if err != nil {
+		return err
+	}
+
+	for _, name := range ifnames {
+		// DHCP lease
+		client, err := dhclient.New(dhclient.V4, name, 30*time.Second, 5)
```
Comment (GanShun) [inline cmds/pxeboot/pxeboot.go]: If you specify V4 here, don't you get a client who's solicit will use the V4 solicit? and packet.Boot right now doesn't support V4 right?

File: pkg/dhclient/dhclient.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+			retry:    retry,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("invalid DHCP version: %d", version)
+	}
+}
+
+func (c *Client) Solicit() (Packet, error) {
+	if c.d4client != nil {
```
Comment (rminnich) [inline pkg/dhclient/dhclient.go]: This seems kinda awkward to me. I understand why you do it this way, but ...

File: pkg/dhclient/dhclient.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	}
+
+	return getPacket4(func() (bool, dhcp4.Packet, error) {
+		return client.Renew(pkg.p)
+	}, retry)
+}
+
+func getPacket4(f func() (bool, dhcp4.Packet, error), retry int) (Packet, error) {
+	for i := 0; i < retry || retry < 0; i++ {
+		success, packet, err := f()
```
Comment (rminnich) [inline pkg/dhclient/dhclient.go]: here's a perfect place to stash statistics in a client. But we would need to either return said statistics or pass the client in.

File: pkg/dhclient/dhclient.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+	// Gateway is the gateway server, if specified, otherwise nil.
+	Gateway() net.IP
+
+	// DNS are DNS server addresses, if specified, otherwise nil.
+	DNS() []net.IP
+
+	// Boot returns the boot URI and boot parameters if specified,
+	// otherwise an error.
+	Boot() (url.URL, []string, error)
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/dhclient/dhclient.go]: or the Stringer you will write could print the stats :-)

File: .gitignore [hunk 5]
```diff
@@ -47,3 +47,17 @@ vmlinux
 # Python Virtual Env
 *.env
 *.venv
+
+### Vim ###
+# swap
```
Comment (bkochendorfer) [inline .gitignore]: This is fine but you might want to set vim to save swap in `/tmp` or something instead. That way you don't have to worry about this kind of thing in every repo.

**Comment by hugelgupf** [comment]
You'll notice that this actually needs https://github.com/u-root/dhcp4/pull/6/files vendored in. Obv that hasn't happened yet. Please review that first?

**Comment by hugelgupf** [comment]
It's a christmas miracle: a couple thousand lines of code appeared out of nowhere. DHCPv4 works, v6 is still in the works (but we've had it working before, I just rearranged some stuff and broke it...).

pkg/pxe has 1000 lines of tests and 500 lines of code at the moment.

Not ready for review yet; just wanted to post an update :)

**Comment by hugelgupf** [comment]
I'll be splitting this up into multiple PRs to make it easier to review. The pkg/dhclient stuff will be separate at least.

**Comment by hugelgupf** [comment]
This is coming closer to being ready for review. Please start taking a look. There are some TODOs, but some of them may remain for a while :)

File: pkg/dhclient/dhcp4.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
+// Package dhclient provides a unified interface for interfacing with both
```
Comment (rminnich) [inline pkg/dhclient/dhcp4.go]: wrong comment?

File: pkg/pxe/pxe.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+			}
+		}
+	}
+
+	// Go through all labels and download the initrds.
+	for _, label := range c.Entries {
+		// If the initrd was set via the INITRD directive, don't
+		// overwrite that.
+		//
+		// TODO(hugelgupf): Is this really what syslinux does? Does
```
Comment (rminnich) [inline pkg/pxe/pxe.go]: This is why parsing these files is such a mess. The languages are highly improvised.

Comment (hugelgupf) [reply pkg/pxe/pxe.go]: Some day, when this doesn't work, we can go read ipxe source code or whatever. Until then, this is gonna be it :)

**Comment by hugelgupf** [comment]
There's more to be done here, but I'm going to leave that to a different PR and merge this.

**Pull request merged**